File: algorithms/565. Array Nesting/solution.py
class Solution(object):
    def arrayNesting(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        v = [0] * len(nums)
        maxi = cur = 0
        for i in xrange(len(nums)):
            if not v[i]:
                cur = 0
                while not v[i]:
                    v[i] = 1
                    i = nums[i]
                    cur += 1
                maxi = max(maxi, cur)
        return maxi

File: algorithms/565. Array Nesting/solution.cc
class Solution {
public:
    int arrayNesting(vector<int>& nums) {
        int n = nums.size(), maxlen = 0, cur;
        vector<bool> visited(n, false);
        for ( int i = 0, j; i < n; ++i ) {
            if ( ! visited[i] ) {
                cur = 0;
                j = i;
                while ( ! visited[j] ) {
                    visited[j] = true;
                    j = nums[j];
                    ++cur;
                }
                maxlen = max(maxlen, cur);
            }
        }
        return maxlen;
    }
};
